modules/renter/repair.go

```diff
@@ -150,6 +151,21 @@ func (f *file) chunkHosts(chunk uint64) []modules.NetAddress {
 	return old
 }
 
+// expiringContracts returns the contracts that will expire soon.
+// TODO: what if contract has fully expired?
+func (f *file) expiringContracts(height types.BlockHeight) []fileContract {
+	f.mu.RLock()
+	defer f.mu.RUnlock()
+
+	var expiring []fileContract
+	for _, fc := range f.contracts {
+		if height > fc.WindowStart-renewThreshold {
```

Comment:
Maybe something like:
1. Renew any expiring contracts. If a contract can't be renewed, add it's pieces to the repair set.
2. Check for offline hosts. Add their pieces to the repair set.
3. Check for unuploaded pieces. Add those pieces to the repair set.
4. Repair all the missing pieces.